// perfume.js/dist/lib/reportPerf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.reportPerf = void 0;
var config_1 = require("./config");
var getNavigatorInfo_1 = require("./getNavigatorInfo");
var onVisibilityChange_1 = require("./onVisibilityChange");
var utils_1 = require("./utils");
/**
 * Sends the User timing measure to analyticsTracker
 */
var reportPerf = function (measureName, data, rating, attribution, navigationType) {
    var reportTask = function () {
        if (!config_1.config.analyticsTracker) {
            return;
        }
        // Doesn't send timing when page is hidden
        if (onVisibilityChange_1.visibility.isHidden && !['CLS', 'INP'].includes(measureName)) {
            return;
        }
        // Send metric to custom Analytics service
        config_1.config.analyticsTracker({
            attribution: attribution,
            metricName: measureName,
            data: data,
            navigatorInformation: (0, getNavigatorInfo_1.getNavigatorInfo)(),
            rating: rating,
            navigationType: navigationType,
        });
    };
    // Send CLS and INP metrics immediately,
    // because this metrics are reported when page is hidden or closed
    if (['CLS', 'INP'].includes(measureName)) {
        reportTask();
    }
    else {
        (0, utils_1.pushTask)(reportTask);
    }
};
exports.reportPerf = reportPerf;
