// perfume.js/dist/lib/vitalsScore.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVitalsScore = exports.getRating = exports.webVitalsScore = exports.STEP_THRESHOLDS = void 0;
var types_1 = require("./types");
var rtScore = [100, 200];
var tbtScore = [200, 600];
var ntbtScore = [200, 600];
exports.STEP_THRESHOLDS = (_a = {},
    _a[types_1.IThresholdTier.instant] = {
        vitalsThresholds: [100, 200],
        maxOutlierThreshold: 10000,
    },
    _a[types_1.IThresholdTier.quick] = {
        vitalsThresholds: [200, 500],
        maxOutlierThreshold: 10000,
    },
    _a[types_1.IThresholdTier.moderate] = {
        vitalsThresholds: [500, 1000],
        maxOutlierThreshold: 10000,
    },
    _a[types_1.IThresholdTier.slow] = {
        vitalsThresholds: [1000, 2000],
        maxOutlierThreshold: 10000,
    },
    _a[types_1.IThresholdTier.unavoidable] = {
        vitalsThresholds: [2000, 5000],
        maxOutlierThreshold: 20000,
    },
    _a);
exports.webVitalsScore = {
    RT: rtScore,
    TBT: tbtScore,
    NTBT: ntbtScore,
};
var getRating = function (value, vitalsThresholds) {
    if (value <= vitalsThresholds[0]) {
        return 'good';
    }
    return value <= vitalsThresholds[1] ? 'needsImprovement' : 'poor';
};
exports.getRating = getRating;
var getVitalsScore = function (measureName, value) {
    if (!exports.webVitalsScore[measureName]) {
        return null;
    }
    if (value <= exports.webVitalsScore[measureName][0]) {
        return 'good';
    }
    return value <= exports.webVitalsScore[measureName][1] ? 'needsImprovement' : 'poor';
};
exports.getVitalsScore = getVitalsScore;
