// perfume.js/dist/es/steps/steps.js
export var steps = {
    finalMarkToStepsMap: {},
    startMarkToStepsMap: {},
    active: {},
    /**
      An internal data structure to represent the state of active steps.
      In this data structure, the keys represent the navgiation index, starting at 0 -
      i.e - the initial app launch it's index 0, the first navigation is index 1 and so one.
      The value of each navigation index is a Set of all the first marks for the currently active steps.
      When a start mark is hit - it will be added to this state.
      When a corresponding end mark is hit, the start mark previously added will be removed from this navigation state
      Examples:
      Example of the state when home is loading:
      {
        [0]: ['start_navigate_to_home_screen'] // load home - empty array since home has loaded
      }
      Example of the state during the navigation from home to assets
      {
        [0]: [] // load home - empty array since home has loaded
        [1]: ['start_navigate_to_trade_screen']
      }
     */
    navigationSteps: {},
};
/**
 * This method allows to add new steps by passing the start mark
 */
export var addActiveSteps = function (startMark) {
    var _a;
    var newSteps = (_a = steps.startMarkToStepsMap[startMark]) !== null && _a !== void 0 ? _a : [];
    // adding the new steps to the active map
    Object.keys(newSteps).forEach(function (step) {
        if (steps.active[step]) {
            return;
        }
        else {
            steps.active[step] = true;
        }
    });
};
export var resetNavigationSteps = function () {
    steps.navigationSteps = {};
};
/**
 * Removes one step from active steps
 */
export var removeActiveStep = function (step) {
    delete steps.active[step];
};
export var resetActiveSteps = function () {
    steps.active = {};
};
export var resetSteps = function () {
    // reset all values
    steps.startMarkToStepsMap = {};
    steps.finalMarkToStepsMap = {};
    resetActiveSteps();
    resetNavigationSteps();
};
export var getNavigationState = function () { return steps.navigationSteps; };
export var getStepsFromNavigation = function (navIndex) {
    var _a;
    var navigationSteps = getNavigationState();
    return (_a = navigationSteps[navIndex]) !== null && _a !== void 0 ? _a : {};
};
