// perfume.js/dist/es/reportPerf.js
import { config } from './config';
import { getNavigatorInfo } from './getNavigatorInfo';
import { visibility } from './onVisibilityChange';
import { pushTask } from './utils';
/**
 * Sends the User timing measure to analyticsTracker
 */
export var reportPerf = function (measureName, data, rating, attribution, navigationType) {
    var reportTask = function () {
        if (!config.analyticsTracker) {
            return;
        }
        // Doesn't send timing when page is hidden
        if (visibility.isHidden && !['CLS', 'INP'].includes(measureName)) {
            return;
        }
        // Send metric to custom Analytics service
        config.analyticsTracker({
            attribution: attribution,
            metricName: measureName,
            data: data,
            navigatorInformation: getNavigatorInfo(),
            rating: rating,
            navigationType: navigationType,
        });
    };
    // Send CLS and INP metrics immediately,
    // because this metrics are reported when page is hidden or closed
    if (['CLS', 'INP'].includes(measureName)) {
        reportTask();
    }
    else {
        pushTask(reportTask);
    }
};
