// perfume.js/dist/es/vitalsScore.js
var _a;
import { IThresholdTier } from './types';
var rtScore = [100, 200];
var tbtScore = [200, 600];
var ntbtScore = [200, 600];
export var STEP_THRESHOLDS = (_a = {},
    _a[IThresholdTier.instant] = {
        vitalsThresholds: [100, 200],
        maxOutlierThreshold: 10000,
    },
    _a[IThresholdTier.quick] = {
        vitalsThresholds: [200, 500],
        maxOutlierThreshold: 10000,
    },
    _a[IThresholdTier.moderate] = {
        vitalsThresholds: [500, 1000],
        maxOutlierThreshold: 10000,
    },
    _a[IThresholdTier.slow] = {
        vitalsThresholds: [1000, 2000],
        maxOutlierThreshold: 10000,
    },
    _a[IThresholdTier.unavoidable] = {
        vitalsThresholds: [2000, 5000],
        maxOutlierThreshold: 20000,
    },
    _a);
export var webVitalsScore = {
    RT: rtScore,
    TBT: tbtScore,
    NTBT: ntbtScore,
};
export var getRating = function (value, vitalsThresholds) {
    if (value <= vitalsThresholds[0]) {
        return 'good';
    }
    return value <= vitalsThresholds[1] ? 'needsImprovement' : 'poor';
};
export var getVitalsScore = function (measureName, value) {
    if (!webVitalsScore[measureName]) {
        return null;
    }
    if (value <= webVitalsScore[measureName][0]) {
        return 'good';
    }
    return value <= webVitalsScore[measureName][1] ? 'needsImprovement' : 'poor';
};
